### DISCUSSION TITLE
Add build "buildLibsFromSource" option to the esbuild executor

### DISCUSSION THREAD

**Discussion by b-smets** [nrwl/nx#23370] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

Most of the bundlers in NX support the "buildLibsFromSource" option. The default value for this option is typically `true` which means that whenever the bundler executes, all dependencies are built again from source, even when they are buildable libs. However, setting this to `false` will instruct the bundler to generate a temporary tsConfig file with paths of buildable libs mapped to the build artifacts in the dist folder. The bundler will then no longer rebuild the previously built libs.

## Motivation

Allowing the esbuild executor to also disable the "buildLibsFromSource" option introduces support for incremental builds, similar to the other bundler executors. The major benefit of this is that the libs can be built individually, using their own tsConfig settings.
At this point, the compiler options in the esbuild app need to be at least as lax as the sum of all its dependencies, as it is the app's tsConfig that is used to also build the libs. When consuming a legacy lib in a new app, you now lose all the good stuff of modern compiler options as the legacy lib potentially requires disabling them.

As an example, use the following steps:
1. Generate 2 buildable libs A and B using the standard @nx/js generator.
2. In lib A, explicitly disable `strictNullChecks` and add a nullcheck violation (e.g. `const a: string = null;`)
3. In lib B, explicitly disable `noImplicitAny` and add a any violation (e.g. a function without typed parameters)
4. Generate a @nx/js app using esbuild as the bundler
5. Consume both libs in a source file in the app to introduce the dependencies
6. Build the app

=> The build will fail as the app did not disable `strictNullChecks` and `noImplicitAny`, and the only way out is neutering the compiler for the app, disabling both of these.

## Suggested Implementation

Similar to the other bundlers, I would expect the following behaviour:

- buildLibsFromSource = true => Build all libs again from source, regardless of being buildable, using the esbuild process.
- buildLibsFromSource = false => Use the build artifacts of buildable libs instead of building them again. Non-buildable libs are still to be build from source.
- Omitted => Same behaviour as when set to `true`.

**Comment by pawelofieldly** [comment]
I have the same issue with NX project. We have couple legacy libs.

**Comment by danielsharvey** [comment]
Just noting what I think is a related issue - https://github.com/nrwl/nx/issues/26376
